junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/KafkaClusterExtension.java

```diff
@@ -555,36 +574,45 @@ private void injectFields(ExtensionContext context, Object testInstance, Class<?
                     }
                 });
 
-        findFields(testClass,
-                field -> predicate.test(field) && Admin.class.isAssignableFrom(field.getType()),
+        final Map<Class<?>, List<Field>> fieldsByType = findFields(testClass,
+                field -> predicate.test(field) && isCandidate(field),
                 HierarchyTraversalMode.BOTTOM_UP)
-                .forEach(field -> {
+                .stream()
+                .collect(Collectors.groupingBy(Field::getType));
+
+        injectClient(Admin.class, KafkaClusterExtension::getAdmin, context, testInstance, fieldsByType);
+        injectClient(Producer.class, KafkaClusterExtension::getProducer, context, testInstance, fieldsByType);
+        injectClient(Consumer.class, KafkaClusterExtension::getConsumer, context, testInstance, fieldsByType);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T, X extends T> void injectClient(Class<T> clientType, ClientFactory<T, X> clientFactory, ExtensionContext context, Object testInstance,
+                                                      Map<Class<?>, List<Field>> fieldsByType) {
+        fieldsByType.entrySet().stream()
+                .filter(entry -> clientType.isAssignableFrom(entry.getKey()))
+                .map(Map.Entry::getValue)
+                .flatMap(List::stream)
+                .filter(field -> {
                     try {
-                        makeAccessible(field).set(testInstance, getAdmin(
-                                "field " + field.getName(),
-                                field,
-                                field.getType().asSubclass(Admin.class),
-                                context));
+                        return field.get(testInstance) == null;
```

Comment:
The code now takes an inconsistent approach to visibility.  You need to mutate the visibility before you check for nullness.  Add a test case that checks that private fields are supported too.

```
java.lang.IllegalAccessException: class io.kroxylicious.testing.kafka.junit5ext.KafkaClusterExtension cannot access a member of class io.kroxylicious.testing.kafka.junit5ext.InstanceFieldExtensionTest with modifiers "private"
```

Reply:
Good point. Addressed in e65ca1ba1d2f14e0a2a6e2ff87f57836bf1f890b.